### PR TITLE
Refactor playlist core

### PR DESCRIPTION
**Pull request by sangwoo108**
The goal of this refactoring is
* Clean up data structures
  * Distinguish Playlist and PlaylistItem
  * Currently the core assumes that there's only one playlist
  * But our spec require us to support multiple playlists for users
* Make use of new js script to detect media files
  * Now we don't get separated auido/video files. Thus, cleaned up
    code that we don't need anymore.
  * Temporarily, bundle the script in the binary. this should be
    converted to component extensions.
  * also it'd be good to have unittests for this script.

Resolves https://github.com/brave/brave-browser/issues/23967

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/playlist/playlist_constants.h [hunk 1]
```diff
@@ -8,28 +8,11 @@
...
-    "createParams.playlistThumbnailUrl";
-constexpr char kPlaylistCreateParamsVideoMediaFilesPathKey[] =
-    "createParams.videoMediaFiles";
-constexpr char kPlaylistCreateParamsAudioMediaFilesPathKey[] =
-    "createParams.audioMediaFiles";
+constexpr char kPlaylistItemIDKey[] = "id";
+constexpr char kPlaylistItemThumbnailPathKey[] = "thumbnailPath";
+constexpr char kPlaylistItemMediaFilePathKey[] = "mediaFilePath";
+constexpr char kPlaylistItemReadyKey[] = "ready";
+constexpr char kPlaylistItemTitleKey[] = "titles";
```
Comment (sangwoo108) [inline components/playlist/playlist_constants.h:15]:
```suggestion
constexpr char kPlaylistItemTitleKey[] = "title";
```

File: components/playlist/playlist_download_request_manager.cc [hunk 2]
```diff
@@ -108,91 +117,122 @@ void PlaylistDownloadRequestManager::RunMediaDetector(const std::string& url) {
...
 bool PlaylistDownloadRequestManager::ReadyToRunMediaDetectorScript() const {
-  if (!media_detector_script_.empty() && web_contents_ && web_contents_ready_)
-    return true;
+  return !media_detector_script_.empty();
+}
 
-  return false;
+void PlaylistDownloadRequestManager::DidFinishNavigation(
+    content::NavigationHandle* navigation_handle) {
+  if (!IsYouTubeDomain(web_contents()->GetLastCommittedURL()))
```
Comment (sangwoo108) [inline components/playlist/playlist_download_request_manager.cc:130]: "vimeo" needs this too.

File: components/playlist/playlist_download_request_manager.cc [hunk 3]
```diff
@@ -108,91 +117,122 @@ void PlaylistDownloadRequestManager::RunMediaDetector(const std::string& url) {
...
+      "detected": boolean,
+      "mimeType": "video" | "audio",
+      "name": string,
+      "pageSrc": url,
+      "pageTitle": string
+      "src": url
+      "thumbnail": url | undefined
+    }
+  ]
+
```
Comment (sangwoo108) [inline components/playlist/playlist_download_request_manager.cc:194]:
```suggestion
```

File: components/playlist/playlist_download_request_manager.cc [hunk 4]
```diff
@@ -108,91 +117,122 @@ void PlaylistDownloadRequestManager::RunMediaDetector(const std::string& url) {
   DCHECK_GE(in_progress_youtube_urls_count_, 0);
   in_progress_youtube_urls_count_++;
```
Comment (sangwoo108) [inline components/playlist/playlist_download_request_manager.cc:118]:
```suggestion
  in_progress_urls_count_++;
```

File: components/playlist/playlist_media_file_downloader.cc [hunk 5]
```diff
@@ -47,132 +47,36 @@ net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTagForURLLoad() {
...
           context->content::BrowserContext::GetDefaultStoragePartition()
               ->GetURLLoaderFactoryForBrowserProcess()),
       request_helper_(std::make_unique<api_request_helper::APIRequestHelper>(
           GetNetworkTrafficAnnotationTagForURLLoad(),
           url_loader_factory_)),
-      source_media_files_dir_(source_media_files_dir),
-      unified_media_file_name_(unified_media_file_name),
-      media_file_path_key_(media_file_path_key),
-      create_params_path_key_(create_params_path_key) {}
+      unified_media_file_name_(unified_media_file_name) {}
```
Comment (sangwoo108) [inline components/playlist/playlist_media_file_downloader.cc:63]:
```suggestion
      media_file_name_(media_file_name) {}
```

File: components/playlist/playlist_media_file_downloader.h [hunk 6]
```diff
@@ -103,17 +87,12 @@ class PlaylistMediaFileDownloader {
...
 
-  const base::FilePath::StringType source_media_files_dir_;
   const base::FilePath::StringType unified_media_file_name_;
-  const std::string media_file_path_key_;
-  const std::string create_params_path_key_;
 
   // All below variables are only for playlist creation.
   base::FilePath playlist_dir_path_;
   base::Value current_playlist_;
   std::string current_playlist_id_;
```
Comment (sangwoo108) [inline components/playlist/playlist_media_file_downloader.h:95]: Replace these with PlaylistItem struct

File: components/playlist/playlist_service.cc [hunk 7]
```diff
@@ -195,41 +185,37 @@ void PlaylistService::OnThumbnailDownloaded(const std::string& id,
...
 }
 
 base::Value PlaylistService::GetPlaylistItem(const std::string& id) {
-  const base::Value* item_value_ptr =
-      prefs_->Get(kPlaylistItems)->FindDictKey(id);
-  if (item_value_ptr) {
-    base::Value item = item_value_ptr->Clone();
-    item.RemoveKey(kPlaylistCreateParamsKey);
-    return item;
+  if (const base::Value* item_value_ptr =
```
Comment (sangwoo108) [inline components/playlist/playlist_service.cc:214]: // TODO: get items per |id|

File: components/playlist/playlist_service.cc [hunk 8]
```diff
@@ -249,19 +236,16 @@ void PlaylistService::RecoverPlaylistItem(const std::string& id) {
...
 
   const std::string* thumbnail_path_str =
-      playlist_info->FindStringPath(kPlaylistThumbnailPathKey);
+      playlist_info->FindStringPath(kPlaylistItemThumbnailPathKey);
   const bool has_thumbnail =
       !!thumbnail_path_str && !thumbnail_path_str->empty();
   if (!has_thumbnail)
     DownloadThumbnail(id);
 
   const std::string* video_media_file_path =
```
Comment (sangwoo108) [inline components/playlist/playlist_service.cc:245]:
```suggestion
  const std::string* media_file_path =
```

File: components/playlist/playlist_types.h [hunk 9]
```diff
@@ -11,68 +11,42 @@
...
-// TODO(simonhong): Rename to PlaylistItemInfo
-struct PlaylistInfo {
-  PlaylistInfo();
-  PlaylistInfo(const PlaylistInfo& rhs);
-  PlaylistInfo& operator=(const PlaylistInfo& rhs);
-  PlaylistInfo(PlaylistInfo&& rhs) noexcept;
-  PlaylistInfo& operator=(PlaylistInfo&& rhs) noexcept;
-  ~PlaylistInfo();
+struct PlaylistItemInfo {
+  PlaylistItemInfo();
```
Comment (sangwoo108) [inline components/playlist/playlist_types.h:38]:
```suggestion
  PlaylistItemInfo();
  PlaylistItemInfo(base::Value);
```

File: components/playlist/playlist_service.cc [hunk 10]
```diff
@@ -87,19 +79,19 @@ void PlaylistService::Shutdown() {
   task_runner_.reset();
 }
 
-void PlaylistService::RequestDownload(const std::string& url) {
-  download_request_manager_->GeneratePlaylistCreateParamsForYoutubeURL(url);
+void PlaylistService::RequestDownloadMediasFromPage(const std::string& url) {
+  download_request_manager_->GetMediaFilesFromPage(url);
```
Comment (sangwoo108) [inline components/playlist/playlist_service.cc:83]: playlist should be considered

File: components/playlist/playlist_types.h (hunk 9 above)
Comment (sangwoo108) [reply components/playlist/playlist_types.h:38]: We'll see if this is really useful.

File: components/playlist/playlist_download_request_manager.cc [hunk 11]
```diff
@@ -50,7 +62,11 @@ PlaylistDownloadRequestManager::PlaylistDownloadRequestManager(
       delegate_(delegate),
       media_detector_component_manager_(manager) {
   observed_.Observe(media_detector_component_manager_);
+
   media_detector_script_ = media_detector_component_manager_->script();
+
+  // TODO(sko) This line should be removed when component extension is ready.
```
Comment (simonhong) [inline components/playlist/playlist_download_request_manager.cc:68]: `component extension` is built-in extension (bundled by default). - https://source.chromium.org/chromium/chromium/src/+/main:extensions/docs/component_extensions.md

What we'll use here is `component`.

File: browser/playlist/test/playlist_browsertest.cc [hunk 12]
```diff
@@ -314,45 +289,51 @@ IN_PROC_BROWSER_TEST_F(PlaylistBrowserTest, ApiFunctions) {
...
+  VLOG(2) << "delete item";
   // When a playlist is deleted, we should get 1 notification: deleted.
   ResetStatus();
   service->DeletePlaylistItem(lastly_added_playlist_id_);
   EXPECT_EQ(1, on_playlist_changed_called_count_);
-  EXPECT_TRUE(IsPlaylistChangeTypeCalled(
-      PlaylistChangeParams::ChangeType::kChangeTypeDeleted));
+  CheckIsPlaylistChangeTypeCalled(PlaylistItemChangeParams::Type::kDeleted);
 
+  VLOG(2) << "delete all items";
```
Comment (simonhong) [inline browser/playlist/test/playlist_browsertest.cc:336]: nit: This log should be below at line 342?

File: components/playlist/playlist_types.h [hunk 13]
```diff
@@ -11,68 +11,42 @@
...
-  // TODO(simonhong): Delete this. |create_params| has it.
-  std::string playlist_name;
+  std::string title;
   std::string thumbnail_path;
-  std::string video_media_file_path;
-  std::string audio_media_file_path;
+  std::string media_file_path;
   bool ready{false};
-
-  CreatePlaylistParams create_params;
```
Comment (simonhong) [inline components/playlist/playlist_types.h:75]: Nice to delete `create_params`. So `media_file_path` will have source url till it's ready?

**Review by simonhong — APPROVED** [review]
++ with nits.

File: components/playlist/playlist_download_request_manager.cc (hunk 11 above)
Comment (sangwoo108) [reply components/playlist/playlist_download_request_manager.cc:68]: Thanks for the correction! I thought there's something like 'built-in/bundled' extension haha

File: components/playlist/playlist_types.h (hunk 13 above)
Comment (sangwoo108) [reply components/playlist/playlist_types.h:75]: Yes that's right. also thumbnail_url works like that.

**Pull request merged**